python/ql/lib/semmle/python/frameworks/Asyncpg.qll

```diff
@@ -0,0 +1,192 @@
+/**
+ * Provides classes modeling security-relevant aspects of the `asyncpg` PyPI package.
+ * See https://magicstack.github.io/asyncpg/.
+ */
+
+private import python
+private import semmle.python.dataflow.new.DataFlow
+private import semmle.python.Concepts
+private import semmle.python.ApiGraphs
+
+/** Provides models for the `asyncpg` PyPI package. */
+private module Asyncpg {
+  /** A `ConectionPool` is created when the result of `asyncpg.create_pool()` is awaited. */
+  API::Node connectionPool() {
+    result = API::moduleImport("asyncpg").getMember("create_pool").getReturn().getAwaited()
+  }
+
+  /**
+   * A `Connection` is created when
+   * - the result of `asyncpg.connect()` is awaited.
+   * - the result of calling `aquire` on a c=`ConnectionPool` is awaited.
+   */
+  API::Node connection() {
+    result = API::moduleImport("asyncpg").getMember("connect").getReturn().getAwaited()
+    or
+    result = connectionPool().getMember("acquire").getReturn().getAwaited()
+  }
+
+  /** Reverse lookup of the query argument name for a query method. */
+  private string queryMethodName(string queryArg) {
+    result in ["copy_from_query", "execute", "fetch", "fetchrow", "fetchval"] and
+    queryArg = "query"
+    or
+    result = "executemany" and
+    queryArg = "command"
+  }
+
+  /** `Connection`s and `ConnectionPool`s provide some methods that execute SQL. */
+  class SqlExecutionOnConnection extends SqlExecution::Range, DataFlow::MethodCallNode {
+    string queryArg;
+
+    SqlExecutionOnConnection() {
+      this.calls([connectionPool().getAUse(), connection().getAUse()], queryMethodName(queryArg))
+    }
+
+    override DataFlow::Node getSql() { result in [this.getArg(0), this.getArgByName(queryArg)] }
+  }
+
+  /** Reverse lokup of the path argument name for a method accessing the file system. */
+  private string fileAccessMethodName(string pathArg) {
+    result in ["copy_from_query", "copy_from_table"] and
+    pathArg = "output"
+    or
+    result = "copy_to_table" and
+    pathArg = "source"
+  }
+
+  /** `Connection`s and `ConnectionPool`s provide some methods that access the file system. */
+  class FileAccessOnConnection extends FileSystemAccess::Range, DataFlow::MethodCallNode {
+    string pathArg;
+
+    FileAccessOnConnection() {
+      this.calls([connectionPool().getAUse(), connection().getAUse()], fileAccessMethodName(pathArg))
+    }
+
+    // The path argument is keyword only.
+    override DataFlow::Node getAPathArgument() { result in [this.getArgByName(pathArg)] }
+  }
+
+  /**
+   * Holds if `result` is the result of awaiting `awaitedValue`.
+   */
```

Comment:
I think it's important to highlight _why_ we need this predicate (as discussed offline).

```suggestion
  /**
   * Holds if `result` is the result of awaiting `awaitedValue`.
   *
   * Internal helper predicate to achieve the same as `.awaited()` does for API graphs,
   * but sutiable for use with type-tracking.
   */
```

I'm also wondering if there is any reason it's not called `getAwaited`? -- such that it makes our QL style guide for predicates with results.

AHA! And since this predicate seems to have been added to `ApiGraphs.qll`, should we maybe make perform some internal import of that predicate instead to avoid code duplication?